cartography/intel/aws/iam.py

```diff
@@ -364,38 +385,198 @@ def load_user_access_keys(neo4j_session, user_access_keys, aws_update_tag):
 
 
+def ensure_list(obj):
+    if not isinstance(obj, list):
+        obj = [obj]
+    return obj
+
+
+def _transform_policy_statements(statements, policy_id):
+    count = 1
+    if not isinstance(statements, list):
+        statements = [statements]
+    for stmt in statements:
+        if "Sid" not in stmt:
+            statement_id = count
+            count += 1
+        else:
+            statement_id = stmt["Sid"]
+        stmt["id"] = f"{policy_id}/statement/{statement_id}"
+        if "Resource" in stmt:
+            stmt["Resource"] = ensure_list(stmt["Resource"])
+        if "Action" in stmt:
+            stmt["Action"] = ensure_list(stmt["Action"])
+        if "NotAction" in stmt:
+            stmt["NotAction"] = ensure_list(stmt["NotAction"])
+        if "NotResource" in stmt:
+            stmt["NotResource"] = ensure_list(stmt["NotResource"])
+        if "Condition" in stmt:
+            stmt["Condition"] = json.dumps(ensure_list(stmt["Condition"]))
+    return statements
+
+
+@timeit
+def load_policy(neo4j_session, policy_id, policy_name, policy_type, principal_arn, aws_update_tag):
+    injest_policy = """
+    MERGE (policy:AWSPolicy{id: {PolicyId}})
+    ON CREATE SET
+    policy.firstseen = timestamp(),
+    policy.type = {PolicyType},
+    policy.name = {PolicyName}
+    SET
+    policy.lastupdated = {aws_update_tag}
+    WITH policy
+    MATCH (principal:AWSPrincipal{arn: {PrincipalArn}})
+    MERGE (policy) <-[r:POLICY]-(principal)
+    SET r.lastupdated = {aws_update_tag}
+    """
+    neo4j_session.run(
+        injest_policy,
+        PolicyId=policy_id,
+        PolicyName=policy_name,
+        PolicyType=policy_type,
+        PrincipalArn=principal_arn,
+        aws_update_tag=aws_update_tag,
+    ).consume()
+
+
+@timeit
+def load_policy_statements(neo4j_session, policy_id, policy_name, statements, aws_update_tag):
+    injest_policy_statement = """
+        MATCH (policy:AWSPolicy{id: {PolicyId}})
+        WITH policy
+        UNWIND {Statements} as statement_data
+        MERGE (statement:AWSPolicyStatement{id: statement_data.id})
+        SET
+        statement.effect = statement_data.Effect,
+        statement.action = statement_data.Action,
+        statement.notaction = statement_data.NotAction,
+        statement.resource = statement_data.Resource,
+        statement.notresource = statement_data.NotResource,
+        statement.condition = statement_data.Condition,
+        statement.sid = statement_data.Sid,
+        statement.lastupdated = {aws_update_tag}
+        MERGE (policy)-[r:STATEMENT]->(statement)
+        ON CREATE SET r.firstseen = timestamp()
+        SET r.lastupdated = {aws_update_tag}
+        """
+    neo4j_session.run(
+        injest_policy_statement,
+        PolicyId=policy_id,
+        PolicyName=policy_name,
+        Statements=statements,
+        aws_update_tag=aws_update_tag,
+    ).consume()
+
+
+@timeit
+def load_policy_data(neo4j_session, policy_map, aws_update_tag):
```

Comment:
I will clean it up

Reply:
I want to get away from thinking of roles, groups and users seperatly and instead think of them as AWSPrincipals. So while the upstream requires separate methods when it get to loading it they are all just policies attached to AWSPrincipals. You are going to end up with identical functions.
I think your concerns could be mitigated with some data checking in the load_policy function so it will fail fast on developer machines. However splitting them up means developers would make changes to role inline policies and not group inline policies which is not what I want.

Reply:
My main concern is to make our call graph as simple as possible. I think each object should get its own sync function and each sync function should have its own get, transform, and loads.

> However splitting them up means developers would make changes to role inline policies and not group inline policies which is not what I want.

If we don't want developers to think of this as separate things, then it sounds like we might want to do some consolidation and combine these functions together as one.

Reply:
Fixed